### PR TITLE
DX-314 Mock Controllers

### PR DESCRIPTION
**Pull request by seanbocz**
Tests that did not need to call the APIs in order to serve their function were adjusted to use mock controllers populated with canned responses in exchange for the real Controllers:

Changed mapper structure to take in the controllers it will use. This allows the ability to pass in Mock controllers for testing purposes. 
Added interfaces for all Controllers. 
Adjusted mappers to use interfaces for controller types required.
Added mock controllers for mapper tests. Made a mock product for tests also.
Removed some tests that were excessively trivial. Adjusted tests to work with mock objects.
Removed unneeded variables in mapper tests due to using mock data.

Time for testing mappers reduced from approximately 1 minute 5 seconds to approximately 3 seconds.
@danw-iq @melissakendall

### CODE REVIEW

File: App.cs [hunk 1]
```diff
@@ -35,15 +39,34 @@ public static void Main()
...
-			_pricingMapper = new PricingMapper(_cachedMagentoAuthToken, _cachedEaAuthToken);
-			_orderMapper = new OrderMapper(_cachedMagentoAuthToken, _cachedEaAuthToken);
-			_entityMapper = new EntityMapper(_cachedMagentoAuthToken, _cachedEaAuthToken);
-			_customerMapper = new CustomerMapper(_cachedMagentoAuthToken, _cachedEaAuthToken);
+			ControllerFactory controllerFactory = new ControllerFactory(_cachedMagentoAuthToken, _cachedEaAuthToken);
+
+			var assetsController = controllerFactory.CreateController(ControllerType.Assets) as AssetsController;
+			var availabilityController = controllerFactory.CreateController(ControllerType.Availability) as AvailabilityController;
+			var catalogsController = controllerFactory.CreateController(ControllerType.Catalogs) as CatalogsController;
+			var classificationContoller = controllerFactory.CreateController(ControllerType.ClassificationTree) as ClassificationTreeController;
```
Comment (melissakendall) [inline App.cs]: unused controller?

File: App.cs [hunk 2]
```diff
@@ -35,15 +39,34 @@ public static void Main()
...
+			var entitiesController = controllerFactory.CreateController(ControllerType.Entities) as EntitiesController;
+			var fieldController = controllerFactory.CreateController(ControllerType.FieldDefinition) as FieldDefinitionController;
+			var ordersController = controllerFactory.CreateController(ControllerType.Orders) as OrdersController;
+			var pricingController = controllerFactory.CreateController(ControllerType.Pricing) as PricingController;
+			var productLibraryController = controllerFactory.CreateController(ControllerType.ProductLibrary) as ProductLibraryController;
+
+			var cartController = controllerFactory.CreateController(ControllerType.Cart) as CartController;
+			var categoryController = controllerFactory.CreateController(ControllerType.Category) as CategoryController;
```
Comment (melissakendall) [inline App.cs]: same here, unused

File: Controllers/ControllerFactory.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		FieldDefinition,
+		Orders,
+		Pricing,
+		ProductLibrary,
+		Cart,
+		Category,
+		CustomAttributes,
+		Customer,
+		Product,
+		Region
```
Comment (melissakendall) [inline Controllers/ControllerFactory.cs]: (nitpick) alphabetical please :p

**Comment by melissakendall** [comment]
Can we move the `IxxxxController` files to a separate folder, perhaps "Controllers->EndlessAisle->Interfaces"? (and same w/ Magento). It makes it easier to navigate

File: Mappers/PricingMapper.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ public void UpsertPricingForCatalogItem(string catalogItemId, decimal price)
 			if (catalogItemId == null)
 				throw new Exception("A catalogItemId is required to upsert pricing.");
 
-			_eaPricingController.CreatePricingResourceItem(new PricingResource(), catalogItemId, price);
+			_eaPricingController.CreatePricingResourceItem(new PricingResource(), catalogItemId, price, false);
```
Comment (melissakendall) [inline Mappers/PricingMapper.cs]: is "false" necessary? the Interface sets it to false

File: Mappers/ProductMapper.cs [hunk 5]
```diff
@@ -166,7 +167,11 @@ public int UpsertMasterProduct(string productMapping, int eaCategoryId, int eaCl
 		 * @return  string          Identifier of a created CatalogItem
 		 */
 		public string AddProductToEndlessAisle(string slug)
-		{        
+		{
+			if (!Regex.IsMatch(slug, RegexPatterns.SlugPattern))
+			{
+				throw new Exception(string.Format("\"{0}\" is in an invalid slug format.", slug));
+			}
```
Comment (melissakendall) [inline Mappers/ProductMapper.cs]: for consistency - pick a style for one line IF statements - brackets (like this) or no brackets (like the change below this one). Doesn't matter to me personally

**Comment by melissakendall** [comment]
Strangely, when I run the app it errors out in Magento authenticate, come see me later and I will show you =P

File: Tests/Tests/Controllers/Magento/CartControllerTests.cs [hunk 6]
```diff
@@ -8,21 +8,21 @@
...
-        //IMPORTANT: Before you can run these tests, ensure the values below are replaced with ones from your Magento system
-        private const int CartId = 3;
-        private const string ItemSku = "Configurable Product";
-        private const string PaymentMethod = "checkmo";
-        private const string ShippingMethod = "flatrate";
-        private const int CustomerId = 2;
-        private const int SaskRegionId = 77;
-
-        private CartController _cartController;
+		//IMPORTANT: Before you can run these tests, ensure the values below are replaced with ones from your Magento system
```
Comment (melissakendall) [inline Tests/Tests/Controllers/Magento/CartControllerTests.cs]: if we are not using real API calls now - do we still need to include these comments?

File: Tests/Tests/MockObjects/Controllers/EndlessAisle/MockAssetsController.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+				MimeType = "image/jpeg",
+				Name = "TestImage.jpg"
+			};
+		}
+
+		public string AuthToken
+		{
+			get
+			{
+				return "WHdvc0h9b3NYQW9zWHZvc1h2JSYBAyc_PhRWBSAYLUFsIBkSMEA9MRYVWAcsPCQQaD8NJiBPLSUeFVgXGjEBJA4bX0EaBBtAOR8I";
```
Comment (melissakendall) [inline Tests/Tests/MockObjects/Controllers/EndlessAisle/MockAssetsController.cs]: if you are using this in multiple places, maybe extract out to base controller

File: Mappers/PricingMapper.cs (hunk 4 above)
Comment (seanbocz) [inline Mappers/PricingMapper.cs]: Personal preference for me is to not pass in optional parameters unless I know for sure what they change. Then I explicitly do for clarity.

**Pull request merged**